### ISSUE TITLE
FetchStrategy.SUBSELECT Issues

### ISSUE THREAD

**Issue by morphace** [Blazebit/blaze-persistence#1887]
Hi!

I intend to use FetchStrategy.SUBSELECT for certain collections, but it doesn't work as expected.

Standard mapping
==============

Entity: 

```
    @JoinColumn(name = "work_order_id")
    @OneToMany(cascade = [CascadeType.ALL], orphanRemoval = true)
    override var timeRecording: MutableList<TimeRecordingEntity> = mutableListOf(),
```

View:

```
    @get:Mapping(fetch = FetchStrategy.SUBSELECT)
    abstract val timeRecording: MutableList<TimeRecordingView>
```

Correlated
========

Entity:

```
    @ElementCollection
    var successorIds: MutableList<String> = mutableListOf()
```

View:

```
    @get:MappingCorrelatedSimple(
        correlationBasis = "successorIds",
        correlated = WorkOrderEntity::class,
        correlationExpression = "id IN correlationKey",
        fetch = FetchStrategy.SUBSELECT
    )
    abstract val successors: List<WorkOrderSiblingView>
```

I can see that the subqueries are executed properly and return the correct rows.

However the collections (we use Kotlin) stay empty.

Is that supposed to work?
Any idea?

Thanks!

Version: 1.6.11
JPA-Provider: Spring Data JPA
DBMS:  PostgreSQL/CockroachDB
Application Server: Spring Boot

**Comment by beikov** [comment]
Hi there,

this is supposed to work, but note that there is a problem with Hibernate ORM and Kotlin with respect to inheritance and how the `override` in Kotlin behaves.

To get consistent behavior, I would suggest you to use Hibernate ORM property mapping i.e. prefix all `jakarta.persistence` and Hibernate annotations with `get:` e.g. `@get:ElementCollection` etc.

**Comment by morphace** [comment]
Hi Christian,

Thanks for the hint.

I have to say that we're still on Hibernate 5.x... but I'll try.

**Comment by morphace** [comment]
Hi,

I followed your advice (set @Access(AccessType.PROPERTY) and add @get: on the entity), but the collections keep being empty...

Any other idea?

**Comment by beikov** [comment]
Can you share the entity model here? I feel like there is a simple little mistake somewhere which is easy to spot when I see the model.

Overall, the problem is that Kotlin creates a new field for every type in a hierarchy that declares a Kotlin property (even the override ones), but places annotations by default on the fields. So Hibernate sees e.g.

```java
class Base {
  @Id
  int id;
  @Column
  String name;

  public int getId() {
    return id;
  }
  public String getName() {
    return name;
  }
}
class Sub extends Base {
  @Column
  String name;

  public String getName() {
    return name;
  }
}
```

Hibernate ORM then decides that `Base#name` is the field to store information for the persistent attribute named `name`, but with Kotlin you can only ever see `null` by calling `Sub#getName()`, because that accesses the `Sub#name` field, which is never written to by Hibernate ORM.

Blaze-Persistence relies on the information that Hibernate ORM provides, so this weird Kotlin behavior results in the same problem there.

**Comment by morphace** [comment]
I need to say that when I use MULTISET instead of SUBSELECT, everything works perfectly (except that CockroachDB handles the subquery inefficitently, because it's performing a FULL SCAN for the subquery, but that's another story and that's I want to switch to SUBSELECT.

My classes:

```

**Comment by morphace** [comment]
As I said - I tried to switch to AccessType.PROPERTY but got the same result when I changed to SUBSELECT.

**Comment by beikov** [comment]
Can you inspect the entity view object that contains the collections to see if there are other same named collections in the object which are properly initialized? I have the feeling that this `abstract val` Kotlin stuff might produce something odd in terms of JVM class files.
Unfortunately, I am no Kotlin expert, so it's hard for me to say what goes wrong here. Have you tried using a Kotlin interface instead yet?

**Comment by morphace** [comment]
Thanks for your support - I'll inspect the objects. I'll check if I can switch to an interface - I'll keep you updated.

**Comment by morphace** [comment]
I've inspected the entity view object - there's no other same named collection.
And I've tried to use an interface instead of an abstract class without success.
I'll try to debug somehow what's happening to hopefully find the reason why it's not working...

**Comment by morphace** [comment]
But somehow it's not really logic - why does it work for MULTISET, but not for SUBSELECT (without any further changes), collections are the same...

**Comment by morphace** [comment]
I've debugged the SQL statements including the bound parameters - interstingly the parameters are set to NULL!

**Comment by beikov** [comment]
>I've debugged the SQL statements including the bound parameters - interstingly the parameters are set to null!

This is a bit more involved then. Could you try to create a reproducer application for this? For example based on one of our [quickstarts](https://github.com/Blazebit/blaze-persistence?tab=readme-ov-file#quickstart)?

**Comment by morphace** [comment]
Ok, I'll try.

Question: could you point me to the spot where those parameters are injected to the query, so I can try to debug myself?

**Comment by beikov** [comment]
The source query is copied in `AbstractCorrelatedSubselectTupleTransformer#prepare` to then construct the subselect query.

**Comment by morphace** [comment]
Ok, I tried to chase the issue by debugging, which is not easy due to the complexity. But doing that at least I've got a clue where the problem probably is.

What I understood is that BP is taking the root level's query parameters in order to use them for the correlated queries.

However while debugging I could not find, where the values of those parameters would be transferred to the lower levels (AbstractCorrelatedSubselectTupleTransformer). The cb's parameter values are empty by the time of the instantiation of the transformer.

I need to say that I used entity view repositories so far and from what I understood the query method's parameters should be taken and used in the correlated queries. However somehow those values don't make it to the JpaParameters (unlike optional parameters for example).

Consequently, which supports my theory, when I use the criteria builder rather than the entity view repo, the queris function as they are supposed to.

I hope this analysis gives you a hint about what to check?

**Comment by beikov** [comment]
That's very interesting. Thanks for the analysis. I'd still need a simple reproducer for debugging to see where it goes wrong, but with your description, I'm pretty sure this should be easily possible.

In the meantime, I guess you can workaround this by constructing the query "by hand" i.e. with the Blaze-Persistence `CriteriaBuilder` API?

**Comment by morphace** [comment]
Yes - i constructed the queries "by hand" and that makes it work...

**Comment by beikov** [comment]
Thanks for confirming and sorry for the trouble.

**Comment by re1709** [comment]
Is this issue linked to this https://github.com/Blazebit/blaze-persistence/issues/1963

I found when the subselect's in our entity views are changed the issue in 1963 doesn't occur. 

Definitely seems something to do with parameters, as a workaround of setting a named parameter on the builder and using an expression in the builder then works even with subselects present in the entity view.
